src/Microsoft.AspNet.SignalR.Client/Connection.cs

```diff
@@ -469,39 +473,80 @@ protected virtual string OnSending()
         {
-            _connectionData = OnSending();
+            // Captured and used by StartNegotiation to track attempts
+            var negotiationAttempts = 0;
 
-            return transport.Negotiate(this, _connectionData)
-                            .Then(negotiationResponse =>
-                            {
-                                VerifyProtocolVersion(negotiationResponse.ProtocolVersion);
+            Task CompleteNegotiation(NegotiationResponse negotiationResponse)
+            {
+                ConnectionId = negotiationResponse.ConnectionId;
+                ConnectionToken = negotiationResponse.ConnectionToken;
+                _disconnectTimeout = TimeSpan.FromSeconds(negotiationResponse.DisconnectTimeout);
+                _totalTransportConnectTimeout = TransportConnectTimeout + TimeSpan.FromSeconds(negotiationResponse.TransportConnectTimeout);
 
-                                ConnectionId = negotiationResponse.ConnectionId;
-                                ConnectionToken = negotiationResponse.ConnectionToken;
-                                _disconnectTimeout = TimeSpan.FromSeconds(negotiationResponse.DisconnectTimeout);
-                                _totalTransportConnectTimeout = TransportConnectTimeout + TimeSpan.FromSeconds(negotiationResponse.TransportConnectTimeout);
+                // Default the beat interval to be 5 seconds in case keep alive is disabled.
+                var beatInterval = TimeSpan.FromSeconds(5);
 
-                                // Default the beat interval to be 5 seconds in case keep alive is disabled.
-                                var beatInterval = TimeSpan.FromSeconds(5);
+                // If we have a keep alive
+                if (negotiationResponse.KeepAliveTimeout != null)
+                {
+                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
+                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
 
-                                // If we have a keep alive
-                                if (negotiationResponse.KeepAliveTimeout != null)
-                                {
-                                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
-                                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
+                    beatInterval = _keepAliveData.CheckInterval;
+                }
+                else
+                {
+                    _reconnectWindow = _disconnectTimeout;
+                }
+
+                Monitor = new HeartbeatMonitor(this, _stateLock, beatInterval);
+
+                return StartTransport();
+            }
 
-                                    beatInterval = _keepAliveData.CheckInterval;
-                                }
-                                else
+            Task StartNegotiation()
```

Comment:
`await` is for the weak 😸 

(but seriously: I can't change this to async/await unless we drop net40)